### PR TITLE
Aden keras model

### PR DESCRIPTION
**Pull request by hcw1026**

### CODE REVIEW

**Comment by pyg4thr1x** [comment]
Little overview of my commits:
- _Rename model to projectname_: self-explanatory, just moved model.py out to the main folder and renamed it to projectname.py;
- _Tidy up documentation_: just tidied up the docs, no change to code has been made; 
- _Avoid importing everything from Keras..._: @hcw1026, I am sorry but I really couldn't stand three lines of import from keras.layers... I've just substituted tf.keras.layers throughout the code, but feel free to revert this commit if you really really liked your huge import;

- _Update create_config_json_: this is the only actual major change, and here I'd like your opinion. Fist of all, I've changed the output file extension to .json. Most importantly though, my idea would be to have three 'categories of parameters' saved in the json file:
1) the specs for the dataset, things that you shouldn't change unless you rebuild the database and the TFRecords from scratch, such as default number of tags database (without filtering), number of mel_bands and sample rate
2) parameters that specify how the dataset is to be parsed, such as window size, which tags to keep and which tags to merge (will need to add that to the input.py module...)
3) parameters for training, such as learning rate, which loss and which optimizer to use, on and on...

**Comment by pyg4thr1x** [comment]
Further suggested change: I'd completely remove update_config_json. Nobody will use it, just open the file and edit what you want.

But please give me feedback on my commits, especially the change to the json structure

**Comment by pyg4thr1x** [comment]
[config.txt](https://github.com/pukkapies/urop2019/files/3505280/config.txt)

**Comment by callesonne** [comment]
`tags': [
                ['rock', 'pop', 'electronic', 'dance', 'hip-hop', 'jazz', 'metal'],
                ['rock', 'pop', 'electronic', 'dance', 'hip-hop', 'jazz', 'metal', 'male', 'female', 'instrumental'],`
Why is this the default for tags? @pyg4thr1x

**Comment by callesonne** [comment]
And why are some parameters hard coded and some not, for the create_json function?

**Comment by pyg4thr1x** [comment]
> Why is this the default for tags?

That's not the default, my idea was to have a 'presets' list with a bunch of tags/merge_tags training options (maybe the ones where we managed to get higher accuracy?), while still leaving the possibility of total customisation if needed. We are gonna try and train with a bunch of different dataset configurations, right?

**Comment by pyg4thr1x** [comment]
> And why are some parameters hard coded and some not, for the create_json function?

No particular reason...
Originally everything was passed as a parameter to the create_json function, and that clearly becomes unfeasible once you add more and more parameter. I just left the dataset_specs then.

Maybe the best way is to just use a **kwargs keyword

**Comment by callesonne** [comment]
> > And why are some parameters hard coded and some not, for the create_json function?
> 
> No particular reason...
> Originally everything was passed as a parameter to the create_json function, and that clearly becomes unfeasible once you add more and more parameter. I just left the dataset_specs then.
> 
> Maybe the best way is to just use a **kwargs keyword

I think easiest way is to have everything hardcoded like the paper, that makes the code simple and as you said to change it up we will do that manually anyways

**Comment by pyg4thr1x** [comment]
Fixed, take a look now

**Comment by pyg4thr1x** [comment]
All commits after  remove update_json are only doc changes... Sorry for messing up as usual

**Comment by pyg4thr1x** [comment]
Ok, this last commit of mine is _probably_ an exaggeration, but now the config.json file is printed with the preset list elements on the same line (😊)

**Review by callesonne — DISMISSED** [review]
I am happy with this

**Comment by hcw1026** [comment]
line 125, is it a typo? there are two train_options

**Comment by hcw1026** [comment]
and what is pretty print? I assume it only affect how the dictionaries are printed, but won't affect how we input data into training? Other than that I am happy to merge

**Comment by callesonne** [comment]
I agree with Aden.  Also a commit I made this morning was to go from (default) 1024 num_units in the dense layer to 500 as used in the paper

**Comment by callesonne** [comment]
I actually didnt read through the spec frontend too carefully yesterday... I notice that you use 32 default filters, @hcw1026, whilst they use 16. Is there a reason for this or was it just an error? Similarly with num_units in output layer (backend). You picked 1024 whilst they used 500.

**Comment by hcw1026** [comment]
> I actually didnt read through the spec frontend too carefully yesterday... I notice that you use 32 default filters, @hcw1026, whilst they use 16. Is there a reason for this or was it just an error? Similarly with num_units in output layer (backend). You picked 1024 whilst they used 500.

It's just an error actually. I didn't notice they mentioned it in the paper. Yeah let's set default the same as what the paper used. At the end the numbers will depend on config.json though

**Review by callesonne — DISMISSED** [review]
I approve everything now, I have not read the pretty print part but thats not so important.

**Comment by hcw1026** [comment]
@callesonne we can merge

**Review by callesonne — APPROVED** [review]
great
